### PR TITLE
Update ecg.py

### PR DESCRIPTION
**Pull request by trueZihaoXu**
Added Pan_Tompkins_Plus_Plus, an improved Pan-Tompkins algorithm

### CODE REVIEW

File: biosppy/signals/ecg.py [hunk 1]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
     return utils.ReturnTuple((rpeaks,), ("rpeaks",))
 
 
+def Pan_Tompkins_Plus_Plus_segmenter(self, ecg, fs):
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1384]:
```suggestion
def Pan_Tompkins_Plus_Plus_segmenter(, ecg, fs):
```

File: biosppy/signals/ecg.py [hunk 2]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+    """ ECG QRS-Peak Detection Algorithm 
+
+    Follows the approach by Md Niaz and Naimul [MdNiNai22]
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1387]:
```suggestion
    Follows the approach by Md Niaz and Naimul [MdNiNai22]_.
```

File: biosppy/signals/ecg.py [hunk 3]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+
+    Parameters
+    ----------
+    
+    ecg : array
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1392]:
```suggestion
    signal : array
```

File: biosppy/signals/ecg.py [hunk 4]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+        Input raw ECG signal.
+    fs: int, float, optional
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1394]:
```suggestion
    sampling_rate: int, float, optional
```

File: biosppy/signals/ecg.py [hunk 5]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+    ''' Noise Cancelation (Filtering) (5-18 Hz) '''
+
+    if fs == 200:
+        ''' Remove the mean of Signal '''
+        #If fs=200 keep frwquency 5-12 Hz otherwise 5-18 Hz
+        ecg = ecg - np.mean(ecg)  
+
+        Wn = 12*2/fs
+        N = 3
+        a, b = signal.butter(N, Wn, btype='lowpass')
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1430]:
```suggestion
        a, b = ss.butter(N, Wn, btype='lowpass')
```

Is this function from scipy? If so, the name of the library in this file was imported as ss :)

File: biosppy/signals/ecg.py [hunk 6]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+        ecg_l = signal.filtfilt(a, b, ecg)
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1431]:
```suggestion
        ecg_l = ss.filtfilt(a, b, ecg)
```

File: biosppy/signals/ecg.py [hunk 7]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+    m_selected_RR = 0
+    mean_RR = 0
+    ser_back = 0
+
+
+    ''' Noise Cancelation (Filtering) (5-18 Hz) '''
+
+    if fs == 200:
+        ''' Remove the mean of Signal '''
+        #If fs=200 keep frwquency 5-12 Hz otherwise 5-18 Hz
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1425]:
```suggestion
        #If fs=200 keep frequency 5-12 Hz otherwise 5-18 Hz
```

File: biosppy/signals/ecg.py [hunk 8]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+        ecg = ecg - np.mean(ecg)  
+
+        Wn = 12*2/fs
+        N = 3
+        a, b = signal.butter(N, Wn, btype='lowpass')
+        ecg_l = signal.filtfilt(a, b, ecg)
+        
+        ecg_l = ecg_l/np.max(np.abs(ecg_l)) #Normalize by dividing high value. That reduce time of calculation
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1433]:
```suggestion
        ecg_l = ecg_l/np.max(np.abs(ecg_l)) #Normalize by dividing high value. This reduces time of calculation
```

File: biosppy/signals/ecg.py [hunk 9]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+
+        Wn = 5*2/fs
+        N = 3                                           # Order of 3 less processing
+        a, b = signal.butter(N, Wn, btype='highpass')             # Bandpass filtering
+        ecg_h = signal.filtfilt(a, b, ecg_l, padlen=3*(max(len(a), len(b))-1))
+        ecg_h = ecg_h/np.max(np.abs(ecg_h))  #Normalize by dividing high value. That reduce time of calculation
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1439]:
```suggestion
        ecg_h = ecg_h/np.max(np.abs(ecg_h))  #Normalize by dividing high value. This reduces time of calculation
```

File: biosppy/signals/ecg.py [hunk 10]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+
+    else:
+        ''' Band Pass Filter for noise cancelation of other sampling frequencies (Filtering)'''
+        f1 = 5 #3 #5                                          # cutoff low frequency to get rid of baseline wander
+        f2 = 18 #25  #15                                         # cutoff frequency to discard high frequency noise
+        Wn = [f1*2/fs, f2*2/fs]                         # cutoff based on fs
+        N = 3                                           
+        a, b = signal.butter(N=N, Wn=Wn, btype='bandpass')   # Bandpass filtering
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1447]:
```suggestion
        a, b = ss.butter(N=N, Wn=Wn, btype='bandpass')   # Bandpass filtering
```

File: biosppy/signals/ecg.py [hunk 11]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+        ecg_h = signal.filtfilt(a, b, ecg, padlen=3*(max(len(a), len(b)) - 1))
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1448]:
```suggestion
        ecg_h = ss.filtfilt(a, b, ecg, padlen=3*(max(len(a), len(b)) - 1))
```

File: biosppy/signals/ecg.py [hunk 12]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+                    
+        ecg_h = ecg_h/np.max(np.abs(ecg_h))
+
+    vector = [1, 2, 0, -2, -1]
+    if fs != 200:
+        int_c = 160/fs
+        b = interp1d(range(1, 6), [i*fs/8 for i in vector])(np.arange(1, 5.1, int_c))
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1455]: You need to import the library in the ecg.py file! If it is the interpolate library from scipy I suggest you add this to the beginning of the file

```
        import scipy.interpolate as interp
```

And then change this line to:

```suggestion
        b = interp.interp1d(range(1, 6), [i*fs/8 for i in vector])(np.arange(1, 5.1, int_c))  
```

File: biosppy/signals/ecg.py [hunk 13]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+
+    vector = [1, 2, 0, -2, -1]
+    if fs != 200:
+        int_c = 160/fs
+        b = interp1d(range(1, 6), [i*fs/8 for i in vector])(np.arange(1, 5.1, int_c))  
+                                                                                    
+    else:
+        b = [i*fs/8 for i in vector]      
+
+    ecg_d = signal.filtfilt(b, 1, ecg_h, padlen=3*(max(len(a), len(b)) - 1))
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1460]:
```suggestion
    ecg_d = ss.filtfilt(b, 1, ecg_h, padlen=3*(max(len(a), len(b)) - 1))
```

File: biosppy/signals/ecg.py [hunk 14]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+    ecg_d = ecg_d/np.max(ecg_d)
+
+
+    ''' Squaring nonlinearly enhance the dominant peaks '''
+
+    ecg_s = ecg_d**2
+    
+    #Smooting
+    sm_size = int(0.06 * fs)       
+    ecg_s = smoother(signal=ecg_s, kernel='flattop', size=sm_size, mirror=True)
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1471]: This is the biosppy function from biosppy.signals.tools, right? It is imported as "st"

```suggestion
    ecg_s = st.smoother(signal=ecg_s, kernel='flattop', size=sm_size, mirror=True)
```

File: biosppy/signals/ecg.py [hunk 15]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+    ecg_d = signal.filtfilt(b, 1, ecg_h, padlen=3*(max(len(a), len(b)) - 1))
+
+    ecg_d = ecg_d/np.max(ecg_d)
+
+
+    ''' Squaring nonlinearly enhance the dominant peaks '''
+
+    ecg_s = ecg_d**2
+    
+    #Smooting
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1469]:
```suggestion
    #Smoothing
```

File: biosppy/signals/ecg.py [hunk 16]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+
+    temp_vector = np.ones((1, round(0.150*fs)))/round(0.150*fs) # 150ms moving window, widest possible QRS width
+    temp_vector = temp_vector.flatten()
+    ecg_m = np.convolve(ecg_s, temp_vector)  #Convolution signal and moving window sample
+
+    delay = delay + round(0.150*fs)/2
+
+
+    pks = []
+    locs = peakutils.indexes(y=ecg_m, thres=0, min_dist=round(0.231*fs))  #Find all the peaks apart from previous peak 231ms, peak indices
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1482]: Please confirm if this the PeakUtils package, so it can be added to the optional requirements for Biosppy :)

Comment (sofia3ms) [reply biosppy/signals/ecg.py:1482]: Also don't forget to import the library at the beginning of the file!

File: biosppy/signals/ecg.py [hunk 17]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+            s12[d - padding] = min([drrs_pad[d - 1], drrs_pad[d + 1]])
+
+    s22 = np.zeros(len(drrs))
+    for d in range(padding, padding + len(drrs)):
+        if drrs_pad[d] >= 0:
+            s22[d - padding] = min([drrs_pad[d + 1], drrs_pad[d + 2]])
+        elif drrs_pad[d] < 0:
+            s22[d - padding] = max([drrs_pad[d + 1], drrs_pad[d + 2]])
+
+    medrr = medfilt(rr, medfilt_order)
```
Comment (sofia3ms) [inline biosppy/signals/ecg.py:1850]:
```suggestion
    medrr = ss.medfilt(rr, medfilt_order)
```

**Review by sofia3ms — CHANGES_REQUESTED** [review]
Hello Zihao, and thank you for your contribution! I suggested some changes that need to be addressed before testing the code with the rest of Biosppy. They are mainly related with imports and consistency with the rest of the library. Let me know if you agree with them

File: biosppy/signals/ecg.py (hunk 5 above)
Comment (trueZihaoXu) [reply biosppy/signals/ecg.py:1430]: yes, it is from scipy. thanks!

File: biosppy/signals/ecg.py (hunk 12 above)
Comment (trueZihaoXu) [reply biosppy/signals/ecg.py:1455]: Yes I agree with you. Thanks

File: biosppy/signals/ecg.py (hunk 16 above)
Comment (trueZihaoXu) [reply biosppy/signals/ecg.py:1482]: the peakutils is this one: https://peakutils.readthedocs.io/en/latest/. Sorry for the missing import

File: biosppy/signals/ecg.py [hunk 18]
```diff
@@ -1381,6 +1381,858 @@ def ASI_segmenter(signal=None, sampling_rate=1000.0, Pth=5.0):
...
+            Empirically obtaind constant used in threshold calculation.
+        ww: int
+            Window width in ms.
+
+    Returns
+    -------
+        th: float
+            Threshold.
+    '''
+    df = pd.DataFrame({"signal": np.abs(x)})
```
[the remark this answers is not in the document]
Comment (trueZihaoXu) [reply biosppy/signals/ecg.py:1945]: maybe convert dataframe to numpy array using pandas.DataFrame.to_numpy?

[the remark this answers is not in the document]
Comment (sofia3ms) [reply biosppy/signals/ecg.py:1945]: For now we don't support pandas in biosppy, so anything that requires a pandas import will not work. Is it possible to estimate the threshold without any pandas function?

**Review by sofia3ms — APPROVED** [review]

**Comment by biggates** [comment]
Please note that this PR uses peakutils, which should be included in requirements.

**Pull request merged**